app/src/mouse/pim447_trackball.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2021 Cedric VINCENT
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <drivers/sensor.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(PIM447, CONFIG_SENSOR_LOG_LEVEL);
+
+/*
+ * It feels more natural and more confortable to square the speed
+ * reported by the PIM447 trackball.
+ */
+static int16_t convert_speed(int32_t value)
+{
+    bool negative = false;
+
+    if (value < 0) {
+        negative = true;
+        value = -value;
+    }
+
+    if (value > 11) {
+        value = 11;
+    }
+
+    value = value * value;
+
+    if (negative) {
+        value = -value;
+    }
+
+    return value;
+}
+
+static void thread_code(void *p1, void *p2, void *p3)
+{
+    const struct device *dev;
+    int result;
+
+    /* PIM447 trackball initialization. */
+
+    const char *label = DT_LABEL(DT_INST(0, pimoroni_pim447_trackball));
+    dev = device_get_binding(label);
+    if (dev == NULL) {
+        LOG_ERR("Cannot get PIM447_TRACKBALL device");
+        return;
+    }
+
+    /* Event loop. */
+
+    bool button_press_sent   = false;
+    bool button_release_sent = false;
+
+    while (true) {
```

Comment:
Is this constantly polling the sensor? That would be quite a hit on power draw, and wouldn't take the configurable polling rate into account.

Reply:
It is polling the sensor every 10 ms; see `k_sleep(K_MSEC(10))` at the end of this loop.  Anyway, it tried to write a driver in interrupt-mode using the INT line of this sensor, but this is not possible because:
- this INT signal toggles during I2C transferts: ![](https://wtf.roflcopter.fr/pics/aYryeICY/FBhLlyAE)
- this INT signal is triggered (low level) without any reason a couple of ms after being cleared:![](https://wtf.roflcopter.fr/pics/UagAlMz0/Q8q4TgWL)

The ![official "driver" from Pimoroni](https://github.com/pimoroni/trackball-python) uses this INT signal, but it turns out it uses it the wrong way:

        while not trackball.get_interrupt():
            time.sleep(0.001)
        [...]

        def get_interrupt(self):
            if self._interrupt_pin is not None:
                return GPIO.input(self._interrupt_pin) == 0 # [my comment] poll pin value
            else:                                           # [my comment] or
                value = self.i2c_rdwr([REG_INT], 1)[0]      # [my comment] poll register value
                return value & MSK_INT_TRIGGERED

Maybe I should contact Pimoroni to know if there is a way to get a clean INT signal.

I didn't take into account the power consumption because I only use the USB HID, but you are right: this might not be a good driver for wireless keyboard.

Reply:
I'd appreciate it if you contacted them, but we should still think of something. There's a possibility we can contact Pimoroni, but we're unlikely to get through to the creators of old Thinkpad trackpoints, and the constant polling issue is likely to pop up there too.

Reply:
Ok. I've just asked for support on Pimoroni forum: https://forums.pimoroni.com/t/pim447-trackball-how-to-get-a-clean-int-signal/17968

Reply:
Hi both, really appreciate the works done so far on this trackball implementation. I'm one of those who look forward to see this work properly & get merged to master :)

Regarding the clean INT signal issue, is it a software or hardware limitation of the trackball? 

I'm asking this because I am thinking to build a wireless keyboard (on ZMK obviously) with Pimoroni trackball and OLED. OLED is already something that could easily drain battery, and if the trackball does as well, I may have to re-consider my plan. 

Might worth a shot checking in their unofficial discord (https://discord.gg/hr93ByC) or raising an issue in their Github page (https://github.com/pimoroni/trackball-python).

Reply:
If you're prepared to do something so dirty, you can debounce the interrupt signal with a couple of passive components, keeping the int line from passing the low threshold during i2c noise, and dropping fast enough on a true interrupt event:

Channel 2 here is the treated interrupt signal during i2c:
![DS1054Z_DS1ZA233608483_2022-03-09_21 12 03](https://user-images.githubusercontent.com/1752435/157539045-65cbae5a-51b2-4dd8-bf7c-1d0003a13ac6.png)

And here it is when the pimoroni fires a true interrupt:
![DS1054Z_DS1ZA233608483_2022-03-09_21 15 43](https://user-images.githubusercontent.com/1752435/157539154-230afc77-6117-499a-8441-9f1ff732d045.png)

If that's not too gross for the community (and there's not a whole ton of i2c activity on a given board that might eventually drive the level low enough for a false trigger), then is this enough to progress with a power-friendly version?

Reply:
I've been looking into this INT problem, and at least on the specimen I've got the only problem is the noise during I2C. If you just guard the interrupt with a semaphore that you release when you're done reading the values, the INT noise doesn't really matter.

One problem you'll have then is that the falling edge of the interrupt may happen before you release the semaphore, which would block you from ever discovering that the PIM has data to read. The workaround there is to manually check if the interrupt pin is low after releasing the semaphore and in that case retaking it and processing the interrupt.

The final catch is that you'll have to wait for around 8 microseconds from when you end i2c communication before releasing the semaphore to avoid the final INT noise.